src/classes/BDI_DataImport_API.cls

```diff
@@ -0,0 +1,102 @@
+    	Id apexJobId;
+    	try {
+            // validate settings
+            if (diSettings.Contact_Custom_Unique_ID__c != null) {
+                // make sure c1 and c2 have the appropriately named field
+                if (!UTIL_Describe.isValidField(UTIL_Namespace.StrTokenNSPrefix('DataImport__c'), 'Contact1_' + diSettings.Contact_Custom_Unique_ID__c) ||
+                   !UTIL_Describe.isValidField(UTIL_Namespace.StrTokenNSPrefix('DataImport__c'), 'Contact2_' + diSettings.Contact_Custom_Unique_ID__c)) {
+                    throw(new BDIException(string.format(Label.bdiContactCustomIdError, 
+                        new string[]{diSettings.Contact_Custom_Unique_ID__c, 
+                            'Contact1_' + diSettings.Contact_Custom_Unique_ID__c,
+                            'Contact2_' + diSettings.Contact_Custom_Unique_ID__c})));
+                }
+            }
+            
+            if (diSettings.Account_Custom_Unique_ID__c != null) {
+                // make sure a1 and a2 have the appropriately named field
+                if (!UTIL_Describe.isValidField(UTIL_Namespace.StrTokenNSPrefix('DataImport__c'), 'Account1_' + diSettings.Contact_Custom_Unique_ID__c) ||
+                   !UTIL_Describe.isValidField(UTIL_Namespace.StrTokenNSPrefix('DataImport__c'), 'Account2_' + diSettings.Contact_Custom_Unique_ID__c)) {
+                    throw(new BDIException(string.format(Label.bdiAccountCustomIdError, 
+                        new string[]{diSettings.Account_Custom_Unique_ID__c, 
+                            'Account1_' + diSettings.Account_Custom_Unique_ID__c,
+                            'Account2_' + diSettings.Account_Custom_Unique_ID__c})));
+                }
+            }
+            
+            // save any settings changes.
+            try {
+                upsert diSettings;
+            } catch (Exception ex) {
+                // this is a complete hack to work around the bug
+                // https://github.com/SalesforceFoundation/Cumulus/issues/1094
+                // for reasons I cannot fathom, we get an error trying to upsert the
+                // settings object even though the UTIL_CustomSettingsFacade has already
+                // done an upsert on it.  This failure only occurs if there is no org defined
+                // custom setting instance.  The workaround is to catch the upsert failing, and
+                // just clear the Id and do an insert.
+                diSettings.Id = null;
+                insert diSettings;
+            }
+            
+            BDI_DataImport_BATCH bdi = new BDI_DataImport_BATCH();
+            apexJobId = Database.executeBatch(bdi, integer.valueOf(diSettings.Batch_Size__c));
+            return apexJobId;
+
+        } catch (Exception ex) {
+            // no reason to log the error if we haven't started the batch,
+            // since it is just a validation error that we are displaying on the page.
+            if(apexJobId != null)
```

Comment:
checking apexJobId is incorrect, since it is uninitialized and would never be set if we hit the exception.  the original code set a variable detecting that it had made it thru validation and was calling executeBatch.